text/0037-node-upgrades.md

```diff
@@ -0,0 +1,98 @@
+
+### casper-node-launcher
+
+The following describes the control flow in the launcher:
+
+![control flow in the launcher](images/0037/casper-launcher-flow.png "control flow in the launcher")
+
+### casper-node
+
+On startup, the node needs to decide what action to take depending upon whether it is an initial run (without fast sync
+this will always be treated as running from genesis), a restart after an upgrade, or a restart after an unexpected
+shutdown.  The previous run (if any) is deduced by examining the stored linear chain, specifically by querying the
+highest block, and comparing that to the current and next (if available) activation points.
+
+Note that if a node starts with a chainspec which specifies `protocol.hard_reset = true`, then all blocks created in or
+after the chainspec's activation point are effectively removed from storage, creating a situation as if the node had
+ceased operating at the end of the appropriate era.  This option is seen as a failsafe, to cater for the scenario where
+the network has ceased operating properly mid-era.
+
+The actions a node will take on startup, in the order in which they are tested, are summarized below:
+
+* HB = highest block
+* PV = protocol version
+* AP = activation point
+* current X = value taken from current chainspec
+* next X = value taken from next upgraded chainspec
+
+| Condition                                           | Scenario                                     | Action                                   |
+| :-------------------------------------------------- | :------------------------------------------- | :--------------------------------------- |
+| no blocks in storage, and PV 1.0.0                  | initial run                                  | commit genesis and continue running      |
+| no blocks in storage, and PV not 1.0.0              | invalid initial run                          | exit with code 102 to downgrade          |
+| HB.era + 1 == current AP and HB is switch block     | restart immediately after upgrade            | commit upgrade and continue running      |
+| HB.era + 1 == current AP and HB is not switch block | invalid run - blocks missing from storage    | exit with code 102 to downgrade          |
+| HB.era + 1 < current AP                             | invalid run - blocks missing from storage    | exit with code 102 to downgrade          |
+| HB.PV != current PV                                 | restart after fork                           | treat as hard-reset and continue running |
+| HB.PV == current PV and no next chainspec available | unplanned shutdown with no scheduled upgrade | continue running                         |
+| HB.era + 1 < next AP                                | unplanned shutdown before scheduled upgrade  | continue running                         |
+| HB.PV != next PV                                    | missed an upgrade and forked                 | treat as hard-reset and continue running |
```

Comment:
If `HB.PV >= next PV`, would it be safe to also exit with code 0 and upgrade? Because e.g. we might be running version 1.1.0, even though the blocks on disk already contain all blocks of 1.2.0 and some of 1.3.0?

Reply:
I think the safer option is to assume the blocks for 1.2.0 and 1.3.0 are not valid and should be deleted.  If they _were_ valid blocks, the user has just wasted some time and energy, nothing more.

I'm actually not sure how a user's setup could end up in that sort of a state without some jiggery-pokery.  They'd have to have e.g. tinkered with the launcher's cached state to make it start running node 1.1.0 rather than 1.3.0, or imported the LMDB files with the newer blocks from some other source.